File: app/src/main/java/me/zhanghai/android/files/provider/common/PosixFileAttributes.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.common;

import android.os.Parcelable;

import java.util.Set;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java8.nio.file.attribute.PosixFilePermission;

public interface PosixFileAttributes extends java8.nio.file.attribute.PosixFileAttributes {

    @NonNull
    PosixFileType type();

    @Override
    default boolean isRegularFile() {
        return type() == PosixFileType.REGULAR_FILE;
    }

    @Override
    default boolean isDirectory() {
        return type() == PosixFileType.DIRECTORY;
    }

    @Override
    default boolean isSymbolicLink() {
        return type() == PosixFileType.SYMBOLIC_LINK;
    }

    @Override
    default boolean isOther() {
        return !isRegularFile() && !isDirectory() && !isSymbolicLink();
    }

    @NonNull
    @Override
    Parcelable fileKey();

    @Override
    @Nullable
    PosixUser owner();

    @Override
    @Nullable
    PosixGroup group();

    @Nullable
    Set<PosixFileModeBit> mode();

    @Nullable
    @Override
    default Set<PosixFilePermission> permissions() {
        Set<PosixFileModeBit> mode = mode();
        return mode != null ? PosixFileMode.toPermissions(mode) : null;
    }

    @Nullable
    ByteString seLinuxContext();
}

File: app/src/main/java/me/zhanghai/android/files/provider/common/PosixFileAttributes.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.common

import android.os.Parcelable
import java8.nio.file.attribute.PosixFileAttributes
import java8.nio.file.attribute.PosixFilePermission

interface PosixFileAttributes : PosixFileAttributes {
    fun type(): PosixFileType

    override fun isRegularFile(): Boolean = type() == PosixFileType.REGULAR_FILE

    override fun isDirectory(): Boolean = type() == PosixFileType.DIRECTORY

    override fun isSymbolicLink(): Boolean = type() == PosixFileType.SYMBOLIC_LINK

    override fun isOther(): Boolean = !isRegularFile && !isDirectory && !isSymbolicLink

    override fun fileKey(): Parcelable

    override fun owner(): PosixUser?

    override fun group(): PosixGroup?

    fun mode(): Set<PosixFileModeBit>?

    override fun permissions(): Set<PosixFilePermission>? = mode()?.toPermissions()

    fun seLinuxContext(): ByteString?
}
